File: app/src/main/java/io/mgba/View/Adapters/RecyclerView/Interfaces/BaseAdapter.java
package io.mgba.View.Adapters.RecyclerView.Interfaces;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.google.common.base.Function;
import java.util.List;

public abstract class BaseAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    protected final List<T> items;
    protected final int layout;
    private final Function<View, RecyclerView.ViewHolder> generateViewHolder;

    public BaseAdapter(List<T> items, int layout, Function<View, RecyclerView.ViewHolder> generateViewHolder){
        this.items = items;
        this.layout = layout;
        this.generateViewHolder = generateViewHolder;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View inflatedView = LayoutInflater.from(parent.getContext()).inflate(layout, parent, false);
        return generateViewHolder.apply(inflatedView);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }
}

File: app/src/main/java/io/mgba/base/BaseAdapter.kt
package io.mgba.base

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import io.mgba.widgets.RecyclerViewItem
import kotlin.collections.ArrayList
import org.lucasr.twowayview.listeners.SectionIndexer

abstract class BaseAdapter<T>(var layout: Int) : RecyclerView.Adapter<BaseAdapter.ViewHolder<T>>() where T: RecyclerViewItem{

    var data: ArrayList<T> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<T> {
        return prepareViewHolder(parent, viewType)
    }

    abstract fun prepareViewHolder(parent: ViewGroup, viewType: Int): ViewHolder<T>

    override fun onBindViewHolder(holder: ViewHolder<T>, position: Int) = holder.bind(data[position])

    override fun getItemCount() = data.size

    fun removeItem(item: T) {
        val pos = getItemPosition(item)
        data.remove(item)

        if(pos != -1)
            notifyItemRemoved(pos)
    }

    fun addItem(item: T) {
        data.add(item)
        notifyItemChanged(data.size)
    }

    fun addItems(items: ArrayList<T>) {
        val size = data.size
        data.addAll(items)
        notifyItemRangeChanged(size - 1 , items.size)
    }

    fun changeItems(items: ArrayList<T>) {
        data = items
        notifyDataSetChanged()
    }

    private fun getItemPosition(item: T): Int {
        for ((i, it) in data.withIndex()){
            if(it == item)
                return i
        }

        return -1
    }

    fun removeItems() {
        data.clear()
        notifyDataSetChanged()
    }

    abstract class ViewHolder<T>(itemView: View) : RecyclerView.ViewHolder(itemView) {

        open fun bind(item: T) = with(itemView) {}
    }
}